[Doc] Update metaprogramming doc (#3009)
taichi-dev/taichi: docs/lang/articles/advanced/meta.md

# Metaprogramming

Taichi provides metaprogramming infrastructures. There are many benefits of metaprogramming in Taichi:

- Enabling the development of dimensionality-independent codes, e.g., codes which are
  adaptive for both 2D/3D physical simulations.
- Improving runtime performance by moving computations from runtime to compile time.
- Simplifying the development of Taichi standard library.

> :::note Taichi kernels are **lazily instantiated** and massive computations can

Reviewer: ```suggestion
:::note Taichi kernels are **lazily instantiated** and large amount of computation can
```

be executed at **compile-time**. Every kernel in Taichi is a template kernel,
even if it has no template arguments. :::

## Template metaprogramming

> By using `ti.template()` as a argument type hint, a Taichi field can be passed into a kernel. Template programming also enable the codes to be reused for fields with different shape:

Reviewer: ```suggestion
By using `ti.template()` as a argument type hint, a Taichi field can be passed into a kernel. Template programming also enables the codes to be reused for fields with different shapes:
```

Author: Good catch!

```python {2}
@ti.kernel

> def copy1D(x: ti.template(), y: ti.template()):

Reviewer: Shall we use snake case such as `copy_1D` or `copy_1d`?

Author: Have modified the function names to the snake case.

    for i in x:
        y[i] = x[i]

a = ti.field(ti.f32, 4)
b = ti.field(ti.f32, 4)
c = ti.field(ti.f32, 12)
d = ti.field(ti.f32, 12)

# Pass field a and b as arguments of the kernel `copy1D`:
copy1D(a, b)

# Reuse the kernel for field c and d:
copy1D(c, d)
```

## Dimensionality-independent programming using grouped indices

> Taichi provides `ti.grouped` syntax which supports grouping loop indices into a `ti.Vector`. It enables dimensionality-independent programming, i.e., codes are adaptive to scenarios of different dimensions automatically:

Reviewer: ```suggestion
Taichi provides `ti.grouped` syntax which supports grouping loop indices into a `ti.Vector`.
It enables dimensionality-independent programming, i.e., codes are adaptive to scenarios of
different dimensionalities automatically:
```

```python {3-10,15-16}
@ti.kernel
def copy1D(x: ti.template(), y: ti.template()):
    for i in x:
        y[i] = x[i]

@ti.kernel
def copy2d(x: ti.template(), y: ti.template()):
    for i, j in x:
        y[i, j] = x[i, j]

@ti.kernel
def copy3d(x: ti.template(), y: ti.template()):
    for i, j, k in x:
        y[i, j, k] = x[i, j, k]

> # Kernels listed above can be unified by one kernel using `ti.grouped`:

Reviewer: ```suggestion
# Kernels listed above can be unified into one kernel using `ti.grouped`:
```

## Matrix & vector metadata

> For matrices, `matrix.m` and `matrix.n` returns the number of columns and rows respectively.

Reviewer: ```suggestion
For matrices, `matrix.m` and `matrix.n` returns the number of columns and rows, respectively.
```

> For matrices, `matrix.m` and `matrix.n` returns the number of columns and rows respectively.
> For vectors, they are treated as matrices with one column in Taichi, `vector.n` is the number of elements of the vector.

Reviewer: ```suggestion
For matrices, `matrix.m` and `matrix.n` returns the number of columns and rows, respectively.
For vectors, they are treated as matrices with one column in Taichi, where `vector.n` is the number of elements of the vector.
```

```python {4-5,7-8}
@ti.kernel
def foo():
    matrix = ti.Matrix([[1, 2], [3, 4], [5, 6]])
    print(matrix.n)  # number of row: 3
    print(matrix.m)  # number of column: 2
    vector = ti.Vector([7, 8, 9])
    print(vector.n)  # number of elements: 3
    print(vector.m)  # always equals to 1 for a vector
```

## Compile-time evaluations

Using compile-time evaluation will allow certain computations to be executed when kernels are being instantiating. This helps the compiler to conduct optimization and reduce the
certain computations overhead at runtime:

> - Use `ti.static` for compile-time branching (for those who come from

Reviewer: ```suggestion
- Use `ti.static` for compile-time branching (for those who are familiar with
```

> Using compile-time evaluation will allow certain computations to be executed when kernels are being instantiating. This helps the compiler to conduct optimization and reduce the
> certain computations overhead at runtime:
> 
> - Use `ti.static` for compile-time branching (for those who come from

Reviewer: ```suggestion
Using compile-time evaluation allows for some computation to be executed when kernels are instantiated. This helps the compiler to conduct optimization and reduce
computational overhead at runtime:

- Use `ti.static` for compile-time branching (for those who are familiar with
```
